### PR TITLE
Enable SE SymbolicCheck to return multiple states

### LINKED ISSUE #5460: Enable SE SymbolicCheck to return multiple states
Update `SymbolicCheck` APIs to let return more than one state when convenient for the check.

The single-returning API must remain preserved.

This is a prerequisite for S2222 related updates.

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5460

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -227,6 +227,8 @@ dotnet_naming_style.I_style.capitalization = pascal_case
 # Rules
 # ----------------------------------------------------------------------------------------------------------------------
 
+dotnet_diagnostic.CA1822.severity = warning # Increase visibility for Member 'xxx' does not access instance data and can be marked as static
```
Comment (pavel-mikula-sonarsource) [inline .editorconfig:230]: See the last commit. I tend to miss them often and this will increase the visibility of the suggestion.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 2]
```diff
@@ -141,19 +138,24 @@ context.Operation.Instance.Kind switch
                 _ => context.State
             };
 
-        private SymbolicContext InvokeChecks(SymbolicContext context, Func<SymbolicCheck, Func<SymbolicContext, ProgramState>> checkDelegate)
+        private void InvokeExitReached(ExplodedNode node)
         {
+            var context = new SymbolicContext(symbolicValueCounter, null, node.State);
+            foreach (var check in checks)
+            {
+                check.ExitReached(context);
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:146]: Why does `ExitReached` have a `ProgramState` return value if it is never used?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 3]
```diff
@@ -114,18 +114,15 @@ private IEnumerable<ExplodedNode> ProcessBranching(ExplodedNode node)
...
-            var context = new SymbolicContext(symbolicValueCounter, node.Operation, node.State);
-            context = InvokeChecks(context, x => x.PreProcess);
-            if (context != null)
+            foreach (var preProcessed in InvokeChecks(new(symbolicValueCounter, node.Operation, node.State), x => x.PreProcess))
             {
-                context = EnsureContext(context, ProcessOperation(context));
-                context = InvokeChecks(context, x => x.PostProcess);
-                if (context != null)
+                var processed = EnsureContext(preProcessed, ProcessOperation(preProcessed));
+                foreach (var postProcessed in InvokeChecks(processed, x => x.PostProcess))
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:120]: Can be simplified with` .Select(x => x.State)`

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs [hunk 4]
```diff
@@ -84,7 +84,7 @@ private IEnumerable<ExplodedNode> ProcessBranching(ExplodedNode node)
         {
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:84]: NITPICK: all the constructor invocations for `ExplodedNode` in this method can be simplified to use target typed `new`.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
I left a couple of comments.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:146]: Historical reasons to reuse `InvokeChecks`... I forgot to remove it as it's pointless now

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:120]: I intentionally closed that issue on SC, because it would make the whole method hard to follow due to the `preProcessed`, `processed`, `postProcessed` logical sequence. It would be confusing to change the type for the last one.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:84]: Mosty done in the other PR :)

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs (hunk 2 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:146]: Will you remove it in this PR?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs:146]: I did, but I forgot to push the commit 🤦

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!
